srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+mod mock;
+
+use srml_support::{
+	StorageMap, EnumerableStorageMap, decl_module, decl_storage,
+	traits::WindowLength
+};
+use parity_codec::Codec;
+use sr_primitives::traits::MaybeSerializeDebug;
+use srml_session::SessionIndex;
+
+/// Rolling Window trait
+pub trait Trait: system::Trait {
+	/// Kind to report with window length
+	type Kind: Copy + Clone + Codec + MaybeSerializeDebug + WindowLength<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as RollingWindow {
+		/// Misbehavior reports
+		///
+		/// It stores every unique misbehavior of a kind
+		// TODO [#3149]: optimize how to shrink the window when sessions expire
+		MisbehaviorReports get(misbehavior_reports): linked_map T::Kind => Vec<SessionIndex>;
+
+		/// Bonding Uniqueness
+		///
+		/// Keeps track of uniquely reported misconducts in the entire bonding duration
+		/// which is currently unbounded (insert only)
+		///
+		/// Footprints need to be unique or stash accounts must be banned from joining
+		/// the validator set after been slashed
+		BondingUniqueness get(bonding_uniqueness): linked_map T::Hash => SessionIndex;
+	}
+}
+
+decl_module! {
+	/// Rolling Window module
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {}
+}
+
+/// Trait for getting the number of misbehaviors in the current window
+pub trait GetMisbehaviors<Kind> {
+	/// Get number of misbehavior's in the current window for a kind
+	fn get_misbehaviors(kind: Kind) -> u64;
+}
+
+impl<T: Trait> GetMisbehaviors<T::Kind> for Module<T> {
+	fn get_misbehaviors(kind: T::Kind) -> u64 {
+		<MisbehaviorReports<T>>::get(kind).len() as u64
+	}
+}
+
+/// Trait for reporting misbehavior's
+pub trait MisbehaviorReporter<Kind, Hash> {
+	/// Report misbehavior for a kind
+	fn report_misbehavior(kind: Kind, footprint: Hash, current_session: SessionIndex) -> Result<(), ()>;
+}
+
+impl<T: Trait> MisbehaviorReporter<T::Kind, T::Hash> for Module<T> {
+	fn report_misbehavior(
+		kind: T::Kind,
+		footprint: T::Hash,
+		current_session: SessionIndex,
+	) -> Result<(), ()> {
+		if <BondingUniqueness<T>>::exists(footprint) {
+			Err(())
+		} else {
+			<BondingUniqueness<T>>::insert(footprint, current_session);
+			<MisbehaviorReports<T>>::mutate(kind, |entry| entry.push(current_session));
+			Ok(())
+		}
+	}
+}
+
+impl<T: Trait> srml_session::OnSessionEnding<T::AccountId> for Module<T> {
+	fn on_session_ending(ending: SessionIndex, _applied_at: SessionIndex) -> Option<Vec<T::AccountId>> {
+		for (kind, _) in <MisbehaviorReports<T>>::enumerate() {
+			let window_length = kind.window_length();
+			<MisbehaviorReports<T>>::mutate(kind, |reports| {
+				// it is guaranteed that `reported_session` happened in the same session or before `ending`
+				reports.retain(|reported_session| {
+					let diff = ending.wrapping_sub(*reported_session);
+					diff < *window_length
+				});
+			});
+		}
+
+		// don't provide a new validator set
+		None
```

Comment:
Same as the above comment, I think you can get rid of this, which is kinda weird, using `on_new_session` instead.